Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,18 +384,172 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
+    if (!has_Dirichlet) return;
+
+    amrex::Box domain = Geom(0).Domain();
+    domain.surroundingNodes();
+    const auto lo = lbound(domain);
+    const auto hi = ubound(domain);
+
+    // loop over all mesh refinement levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr = phi[lev]->array(mfi);
+
+            // loop over dimensions
+            for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+                    // check if the boundary in this dimension should be set
+                    if (!dirichlet_flag[idim]) continue;
+
+                    // check if the boundary values are already correct
+                    // otherwise store the domain boundary indices
+                    int idx_lo, idx_hi;
+                    if (idim == 0){
+                        if ((phi_arr(lo.x, 0, 0) == phi_bc_values_lo[idim]) &&
+                            (phi_arr(hi.x, 0, 0) == phi_bc_values_hi[idim])){
```

Comment:
This check won't work in general since the point `phi_arr(lo.x,0,0)` is not necessarily in the current patch. This will produce out of range accesses. Also, now the phi is set to zero before every solve so the BC needs to be applied every step. (I'll submit a PR to change that after this one gets merged.)

Reply:
I keep forgetting about how the domain decomposition works. Do you have a suggestion for how to get around that problem? One thing we could do is check if the current timestep is > 1 AND if the desired boundary potential at this step is equal to what it was at the previous step. But this will give incorrect results until the change is made to not reset phi after every step.

Reply:
I would leave this check out for now. It can be added later as an optimization if the time spent of setting the boundary conditions is not insignificant.

Reply:
Sounds good. I have removed the check and added a comment saying that a similar test can be implemented if it is found to be needed.
Just a heads up, I also had to edit the input file for the space_charge_initialization_2d test since it now failed with the `AMREX_ALWAYS_ASSERT_WITH_MESSAGE` line we added to throw an error when the electrostatic solver is called with boundary conditions other than `periodic` or `pec`.